download_weather_data.py

```
@@ -0,0 +1,286 @@
+        checkpoint_data = {}
+        if checkpoint_file.exists():
+            with open(checkpoint_file, 'r') as f:
+                checkpoint_data = json.load(f)
+        
+        start_location_idx = 0
+        if checkpoint_data and checkpoint_data.get('output_file') == str(output_file):
+            start_location_idx = checkpoint_data.get('current_location_index', 0)
+            logger.info(f"Resuming from checkpoint (location {start_location_idx}/{len(locations)})")
+        
+        if start_location_idx == 0:
+            csv_file = open(output_file, 'w', newline='', encoding='utf-8')
```

Ruff (0.12.2)

Async functions should not open files with blocking methods like `open`

(ASYNC230)